include/sdsl/rank_support_int.hpp

```diff
@@ -0,0 +1,206 @@
+	\author Christopher Pockrandt
+*/
+#ifndef INCLUDED_SDSL_RANK_SUPPORT_INT
+#define INCLUDED_SDSL_RANK_SUPPORT_INT
+
+/** \defgroup rank_support_group Rank Support (RS)
+ * This group contains data structures which support an sdsl::int_vector with the rank method.
+ */
+
+#include "int_vector.hpp"
+#include "uint128_t.hpp"
+
+// TODO: benchmark the use of compiler hints for branch prediction
+#define likely(x)       __builtin_expect((x),1)
+#define unlikely(x)     __builtin_expect((x),0)
+
+//! Namespace for the succinct data structure library.
+namespace sdsl {
+
+//! The base class of classes supporting rank_queries for a sdsl::int_vector in constant time.
+/*!
+*/
+
+constexpr size_t floor_log2(size_t const n)
+{
+    return (n == 1) ? 0 : 1 + floor_log2(n >> 1);
+}
+
+constexpr size_t ceil_log2(size_t const n)
+{
+    return (n == 1) ? 0 : floor_log2(n - 1) + 1;
+}
+
+template <uint8_t alphabet_size>
+class rank_support_int {
+
+public:
+	typedef typename int_vector<>::size_type size_type;
+	typedef typename int_vector<>::value_type value_type;
+
+    static_assert(alphabet_size > 2, "Rank support is only implemented on int_vectors with an alphabet size of > 2.");
+
+protected:
+
+    // Constructs a bit mask with the pattern w of a given length.
+    // It is concatenated until the length of the bitmask reaches max_length.
+    template <typename uintX_t>
+    static constexpr uintX_t bm_rec(const uintX_t w, const uint8_t length, const uint8_t max_length)
+    {
+        return (length >= max_length) ? w : bm_rec(w | (w << length), length << 1, max_length);
+    }
+
+protected:
+	const int_vector<>* m_v; //!< Pointer to the rank supported bit_vector
+    static constexpr uint8_t t_v{alphabet_size};
+    static constexpr uint8_t t_b{ceil_log2(alphabet_size)};
+	static constexpr uint64_t even_mask{bm_rec<uint64_t>(bits::lo_set[t_b], t_b * 2, 64)};
+	static constexpr uint64_t carry_select_mask{bm_rec<uint64_t>(1ULL << t_b, t_b * 2, 64)};
+	uint64_t masks[alphabet_size]; // TODO: make constexpr and remove init()
+
+    // Count how often value v or smaller occurs in the word w.
+    uint64_t set_positions_prefix(const uint64_t w, const value_type v) const
+    {
+        uint64_t const w_even = even_mask & w; // retrieve even positions
+        uint64_t const w_odd = even_mask & (w >> t_b); // retrieve odd positions
+        // every bit that will be set corresponds to an element <= v
+        // because the preset bit to the left in the precomputed bitmask is not eliminated by the carry bit during the subtraction
+        uint64_t res = (masks[v] - w_even) & carry_select_mask;
+        // since alphabet_size is > 2 and an element uses at least 2 bits, we can shift the odd positions by one to the left
+        // and it is guaranteed that when adding both with OR that no bits that are set will overlap.
+        res |= ((masks[v] - w_odd) & carry_select_mask) << 1;
```

Comment:
Together with the paper, I can kinda follow it, but I'm not too sure what happens here :)